### ISSUE TITLE
Spell check - need phrase suggester, not single term

### ISSUE THREAD

**Issue by patrickzurek** [eXtensibleCatalog/Drupal-Toolkit#527]
JIRA issue created by: rcook
Originally opened: 2010-10-28 03:10 PM

Issue body:
 (nt)

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2011-02-17 07:27 PM

Comment body:

Improve the processing of what we have now

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2011-02-17 09:43 PM

Comment body:

Input from Eileen Daly - 

Smarter â€œsearch insteadâ€� engine. I put in â€œJane Austnâ€�  and â€œAustenâ€� didnâ€™t show up as a possible alternate search.

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2011-03-22 05:46 PM

Comment body:

Explore Google like features and maybe a Library of Congress????? (from Summon demo)

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2011-03-22 05:49 PM

Comment body:

Need multiple terms and use Google Suggests that includes a dropdown list.  Find out what is available widget and API wise.

**Comment by patrickzurek** [comment]
JIRA Comment by user: pkiraly
JIRA Timestamp: 2011-03-22 09:12 PM

Comment body:

This is an initial estimation. But I am not sure it could be done without serious implications elsewhere in DT. We can use N-grams (indexing pairs or triplets of words near each other), but it has performance implications. Using an external API would gave us false positive suggesstions ( relevant suggestions, which are not inside our index). Needs a lots of experiences.

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2011-05-10 03:57 PM

Comment body:

Copy of an email Dave sent:

I did some more looking into Summon and found that it works differently than I thought.

It has several distinct features:

1) A "Google Suggest" -like feature that drops down a box of words and phrases that it thinks you might be interested in.  A user has the option of ignoring the options that appear, and just searching for what they typed in, or selecting something from the drop-down list and then searching for that.

2) If a user hover's over an item in the drop-down list of words, a "more information" box pops up.  This information comes from places like Wikipedia (I am guessing).

3) When a user clicks on the search button after typing in a query, Summon then presents results, and in some cases will offer a "Did you mean:" spelling suggestion. 

For #3, I suspect that this can happen in two cases:
- the user types in a misspelled query and ignores the drop-down list of suggestions, and just hits enter.  
- if they pick something from the drop-down that is NOT in the Summon Index (this is the scenario that Peter proposed to avoid with a two step process, but I disagree, I think we should allow the suggestions to include things that are not in our local index).

As a summary, it is desirable for use to have #1, #2, and #3 in our Drupal Toolkit.  As for prioritization, I would say that #1, and #3 are much more important than #2.

**Comment by patrickzurek** [comment]
JIRA Comment by user: pkiraly
JIRA Timestamp: 2011-05-24 04:57 PM

Comment body:

Just a not-to-forget type note: http://google.com/complete/search?output=toolbar&q=almonsd

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2011-07-19 02:39 PM

Comment body:

Please test.

**Comment by patrickzurek** [comment]
JIRA Comment by user: dlindahl
JIRA Timestamp: 2011-07-21 04:26 PM

Comment body:

The phrase-based spelling correction seems to work well.  Two issues I see right now:

1) Searching is very slow.  Seems like 3 to 5 seconds.  Is this caused be the spell checking code, or something else.  I think we need to add some timing code to see where the slowness comes from.

2) There is already an issue for the need to allow you to press "Enter" to do a search, but there is ANOTHER issue.  If you type in a query, and the google suggest box appears underneath, and then you click the "Search" button, the first time you click it, it makes the google suggest box go away, and you have to click it a SECOND time to conduct the search.  Two clicks to conduct one search.

**Comment by patrickzurek** [comment]
JIRA Comment by user: pkiraly
JIRA Timestamp: 2011-07-25 07:22 PM

Comment body:

Dave, you are correct. Saturday I spent a quite amount of time to find, that the problem is with core Drupal. I made a quick fix initially, but hacking the core is not a good solution, so I di not wanted to continue with it. I will forking that part, and put that behaviour entirelly into the module. That's a pain, but at this time I don't see quicker and equally mainatinable solution. (Hacking Drupal core is not maintainable...)
